torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,246 @@
+      (fc2): Linear(in_features=10, out_features=1, bias=True)
+      (relu2): ReLU()
+    )
+    >>> # attaches an optimizer, since parameters can now be used as usual
+    >>> optim = torch.optim.SGD(mlp.parameters(), lr=0.01)
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization when loading an initialized parameter into an uninitialized one.
+    Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one.
+
+    In the example above, if both layers are `LazyLinear` the initialization
+    order is equivalent to the non-lazy one, so weights would be the same if the same
+    random seed is used. However, if only the first layer is lazy, then the initialization
+    of the weights happens in different order. The second layer at creation time, and the
+    first layer is initialized in the dry run, resulting in different weights than the non
+    lazy version of the network.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
+    dummy forward call in the following scenarios.
+
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that accesses Parameters data in their constructor 
+    >>> optim = torch.optim.AdaGrad(lazy_module.parameters())
+
+    * Use :class:`torch.nn.DataParallel` or :class:`torch.nn.DistributedDataParallel`
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(lazy_module)
+
+    * When dealing with non-default module initialization
+    >>> torch.nn.init.uniform_(lazy_module.weight)
+
+    .. warning:
+    Lazy modules are a new feature under heavy development
+    so changes to the API or functionality can happen at any moment.
+    """
+
+    # modules inheriting from this will change their __class__ to the specified
+    # one after they are fully initialized
+    cls_to_become = None
+
+    def __init__(self: _LazyProtocol, *args, **kwargs):
+        # Mypy doesnt like this super call in a mixin
+        super().__init__(*args, **kwargs)  # type: ignore
+        self._load_hook = self._register_load_state_dict_pre_hook(self._lazy_load_hook)
+        self._initialize_hook = self.register_forward_pre_hook(self._infer_parameters)
+        warnings.warn('Lazy modules are a new feature under heavy development '
+                      'so changes to the API or functionality can happen at any moment.')
+
+    def _save_to_state_dict(self: _LazyProtocol, destination, prefix, keep_vars):
+        r"""Saves module state to `destination` dictionary, containing a state
```

Comment:
Can we leave this function undocumented for now? I think it's very likely to change in the future.